File: src/main/java/com/github/kayjamlang/core/containers/Container.java
package com.github.kayjamlang.core.containers;

import com.github.kayjamlang.core.Stmt;
import com.github.kayjamlang.core.expressions.Expression;
import com.github.kayjamlang.core.opcodes.AccessType;

import java.util.ArrayList;
import java.util.List;

/**
 * Expression group
 */
public class Container extends Stmt implements Cloneable {
    /**
     * Expressions in container
     */
    public List<Stmt> children = new ArrayList<>();

    /**
     * Functions in group
     */
    public List<FunctionContainer> functions = new ArrayList<>();

    /**
     * Create group of container
     * @param children Expressions in group
     * @param line Line of container
     */
    public Container(List<Stmt> children, int line) {
        super(line);

        for(Stmt expression: children){
            if(expression instanceof FunctionContainer)
                functions.add((FunctionContainer) expression);
            else this.children.add(expression);
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public Container clone() throws CloneNotSupportedException {
        Container container = (Container) super.clone();
        container.children = (List<Stmt>) ((ArrayList<Stmt>) children).clone();
        container.functions = (List<FunctionContainer>) ((ArrayList<FunctionContainer>) functions).clone();
        return container;
    }
}

File: src/main/scala/com/github/kayjamlang/core/containers/Container.scala
package com.github.kayjamlang.core.containers

import com.github.kayjamlang.core.ArrayList
import com.github.kayjamlang.core.exceptions.ParserException
import com.github.kayjamlang.core.expressions.Expression
import com.github.kayjamlang.core.opcodes.AccessType

import scala.collection.mutable

class Container(var children: ArrayList[Expression], identifier: AccessType, line: Int, x: Boolean) extends Expression(identifier, line) with Cloneable {
  def this(children: ArrayList[Expression], identifier: AccessType, line: Int) = this(children, identifier, line, true)
  def this(children: ArrayList[Expression], line: Int, x: Boolean = true) = this(children, AccessType.NONE, line, x)


  if (x)
    for (expression <- children) {
      expression match {
        case _ => children += expression
      }
    }

  @SuppressWarnings(Array("unchecked"))
  @throws[CloneNotSupportedException]
  override def clone: Container = {
    val container = super.clone.asInstanceOf[Container]
    container.children = children.clone()
    container
  }
}